keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+  - x509
+  - OIDC
+  - Token Webhook
+2. Changes made via the REST API should be active in the order of minutes without
+requiring a restart of the API server
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+1. Changing the internal behavior of the authentication modes (ex: using
+different fields in x509 certificates to build the `user.Info` object)
+2. Restricting the REST API to the exact structure of the CLI flags
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This change aims to add a new API Kubernetes REST API called
+`AuthenticationConfig`.  It is similar to the `ValidatingWebhookConfiguration`
+API in that it will be watched by the API server and used to construct a struct.
+In this case, this struct with be an `authenticator.Request`, i.e. a dynamic
+authenticator.  This authenticator will be unioned with the other authenticators.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wishes to use Dex on an existing cluster.  They follow Dex's instructions
+to deploy the IDP onto the cluster.  They then create a matching
+`AuthenticationConfig` object with the `spec.type` set to `oidc` along with other
+relevant options.  This allows the API server to validate ID tokens issued by Dex.
+The user configures the `kubectl` OIDC integration and uses a Dex provided
+identity to authenticate to the cluster.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TBD.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+#### Restrictions on asserted identities
+
+To prevent confusion with identities that are controlled by Kubernetes, the
+`system:` prefix will be disallowed in the username and groups contained in the
+`user.Info` object.  A disallowed username will cause authentication to fail.
```

Comment:
It would be nice if the credentials for a disallowed username weren't even forwarded to the dynamic authenticator. The only way of making this work that I can think of is if a static authenticator was able to recognize the format of the credential, but it was the wrong credential, then it could shortcircuit-deny the request. WDYT?

EDIT: alternatively, would it be possible to require clients authenticating with a dynamic authenticator to provide a username hint? (e.g. as a request header, only relevant for OIDC & webhook types). This way, the authentication could be skipped if the requested username doesn't match the prefix, and the request would be auto-denied if the authenticated username didn't match.

Reply:
I believe this is the point that @mikedanese brought up in the sig-auth call around "I do not want to forward credentials that were meant for my static webhook to these dynamic webhooks when my static webhook has a transient failure and authentication fell through to the later authenticators."  This is a valid concern and we should try to address it.

> It would be nice if the credentials for a disallowed username weren't even forwarded to the dynamic authenticator.

This does assume that the API server knows in advance what the username will be.  I think we can only know this for sure with certs and SA tokens.

> The only way of making this work that I can think of is if a static authenticator was able to recognize the format of the credential, but it was the wrong credential, then it could shortcircuit-deny the request. WDYT?

The idea of adding short circuit deny logic into authentication makes me very nervous.  I think the authn stack is much easier to reason about because we do not have to worry about "which authenticator owns a credential."  I could also imagine cases where a user has a webhook today that validates tokens that look like service account JWTs.  I do not want to accidentally break them.

> would it be possible to require clients authenticating with a dynamic authenticator to provide a username hint? (e.g. as a request header, only relevant for OIDC & webhook types). This way, the authentication could be skipped if the requested username doesn't match the prefix, and the request would be auto-denied if the authenticated username didn't match.

I think there are some problems with such an approach:

1. The client has to know the username it is trying to authenticate as.  Some environments use UUIDs which make this type of flow very painful.
1. How would the user tell `kubectl` to pass a such a header?  Would we extend the `exec` credential plugin mechanism?
1. If you have more than one webhook configured and they both use opaque tokens, how do you know which webhook to send the token to?

Here is my current thoughts on how we could make this safer:

1. We require every `AuthenticationConfig` object to have a name with the format of `domain:path_segment`.  Something like `company.com:v1`.  We reserve `[*.]k8s:*` for any future Kube usage we come up with.
1. We make no other changes to `x509` or `oidc` because the validation for those authenticators is in-memory of the API server.
1. For `webhook`, we only pass a token to a webhook for validation if it has the prefix `<name_of_authentication_config>::`.  For example, if we see a token `company:v1::D2tIR6OugyAi70do2K90TRL5A` and we have a webhook configured with the name `company:v1`, we would strip the prefix (so that the webhook does not need to know its name) and pass `D2tIR6OugyAi70do2K90TRL5A` to the webhook.  Note that this guarantees that we never pass a standard OIDC token to a webhook.

The above flow requires that a client be given the name of the webhook out of band in some manner and that it concatenate the prefix to the token (this could be done by an `exec` credential provider though we could add direct support to `kubectl` to help facilitate this).

I cannot think of any approach that does not involve the client providing us extra information to make the `webhook` case not leak tokens meant for one webhook to another webhook.  Encoding it directly into the token as a prefix seems like the simplest approach that could be done today via `exec` plugins.  Having to know the (probably static, maybe even well-known like `github.com:kubernetes`) name of the webhook authenticator seems okay to me.

Thoughts?

Reply:
We could also support a simple form of wildcard by allowing the client to send the token as `company.com:*/D2tIR6OugyAi70do2K90TRL5A` as a way of indicating "I am okay with sending this token to all `company.com` authenticators."

Not sure if that is a good idea, but it would be easy to implement.

Reply:
This proposal makes sense to me.

Reply:
> It requires tokens to be provisioned with the prefix though (unless the client adds them). Is that acceptable?

Overall, I think this is acceptable because it could be made transparent to the end-user.

Authentication webhooks and `exec` plugins work together quite nicely to give a seamless experience.  We also do not need to concern ourselves with `oidc` tokens (they just continue to work as they always have and are never sent to the dynamic webhooks).  I see a few ways the webhooks could handle the prefix requirement:

1. The webhook is aware of the prefix it needs to add (i.e. the prefix is static or the webhook knows how it was configured on a given cluster) and does so automatically; no client side changes are required
1. The webhook is unaware of the prefix (it could be an "old" webhook or the name of the webhook is not stable across clusters) but the client adds it automatically via:
   1. Automatic config by the process that creates the kubeconfig file for the end-user
   1. Probing the cluster somehow for the information (i.e. the `exec` plugin knows how to access some public datastore that tells it what prefix to use on a given cluster)
   1. Manual config by the end-user (the user is informed out of band of the prefix information)

Only the last option leaks this implementation detail to end-users.